Allow for more vectorisation options

jeremyh opened the pull request:
~**note**: only an experiment for now: we'll see how it affects things~. Ready to merge, I think, as it allows new options but should not change existing default behaviour.

This avoids a common issue that surprises people: observations with small holes (such as specklings of nodata pixels, as seen in #184) can blow out the CPU usage of geometry calculation.

Valid data is allowed to be bigger than the actual image, and none of our use cases benefit from holes, so why not fill them
by default?

(this makes scipy a core dependency, rather than optional)

## Options Added

This PR makes valid-data calculation method an enum.

1. `thorough`: Vectorize the valid pixel mask as-is. (**the default**)
2. `filled`: Fill holes in the valid pixel mask before vectorizing. (**a potential new default**. but I haven't changed it in this PR to avoid potential compatibility issues.)
        (Much faster than vanilla if there’s many small nodata holes, as they cause many tiny polygons that are later removed by geom simplification. But mildly slower when no holes exist: ~1 sec.)
3. `convex_hull`:  (suggested by Imam)
        Take convex-hull of valid pixel mask before vectorizing.
        This is much slower than filled, but will work in cases where you have a lot of internal geometry that aren’t holes. Such as SLC-Off Landsat 7 data.
        Requires extra ‘scikit-image’ dependency.
4. `bounds` (a "do nothing" option)
        Use the image file bounds, ignoring actual pixel values.


### Remaining Questions

Is an enum an okay API? Maybe they'll want parameters in the future? (eg. "number of pixels to dilate by"). But maybe in those cases we should just let them set their own `Callable` to mess with the mask themselves in more advanced ways?

### Todo

- [ ] ~Test on more of the existing products. Could `filled` be used universally instead of `thorough`?~
      - (Our integration tests are all scaled down tiny images, so aren't a good reflection of impact)
      - *Decided to leave potential default-behaviour change for a future PR*
- [x] Maybe add "geometry calc method" as a config enum? (Possible options: filled-holes, normal, full-image-bounds...)
- [x] Mention in docs the valid-data-methods
- [ ] ~If `method==bounds`, the `note_measurement` methods could avoid reading any pixels at all. Currently, this option just avoids the shape calculation.~
   - *Performance optimisation can be left for a future PR*

jeremyh committed:
Fill geometry holes by default

This avoids a common issue that surprises people: observations
with small holes can blow out the CPU usage of geometry calculation.

Valid data is allowed to be bigger than the actual image, and none
of our use cases benefit from holes in it.

jeremyh committed:
Make valid-data method an enum

jeremyh committed:
Add convex_hull option to valid_data methods

jeremyh committed:
Clarify docs, names of valid data methods

jeremyh committed:
Removed defunct property

ArpitKubadia commented:
Hello,
Has this functionality been added in the latest released version?
For my application, I am using the ValidDataMethod.bounds from this
However the latest official update which adds multiple bands support is also what I need.

Are there any plans to merge these two?
Thanks!!

jeremyh committed:
Keep the old valid-data method for now

To avoid potential compatibility breakages. This
means we can merge this sooner. (and consider changing
the default at a future date).

jeremyh commented:
> Hello, Has this functionality been added in the latest released version? For my application, 

No sorry. I was hesitant to merge it as changing the default option can be dangerous.

I've just changed this PR to use the existing default, so at least we can merge it as-is safely and allow people to try the new options

jeremyh commented:
Released in [version 0.24](https://github.com/opendatacube/eo-datasets/releases/tag/eodatasets3-0.24.0)
